File: ros2_ws/src/home_robot_bt/home_robot_bt/bt_patrol.py
#!/usr/bin/env python3
"""Custom Behavior Tree patrol node.

This implementation owns the patrol behavior tree instead of delegating the
patrol logic to Nav2's BT Navigator. Nav2 is still used as the navigation action
backend, but selection, fallback, recovery and event publication are implemented
as local BT nodes.
"""

from enum import Enum
import math
import os
import random
import signal
import sys
import time

import rclpy
import yaml
from ament_index_python.packages import get_package_share_directory
from geometry_msgs.msg import PoseStamped, Twist
from nav2_msgs.action import NavigateToPose
from nav2_simple_commander.robot_navigator import BasicNavigator, TaskResult
from rclpy.action import ActionClient
from std_msgs.msg import String


stop_patrol = False


class Status(Enum):
    SUCCESS = 'SUCCESS'
    FAILURE = 'FAILURE'
    RUNNING = 'RUNNING'


class Behavior:
    def __init__(self, name: str):
        self.name = name

    def tick(self, blackboard: dict) -> Status:
        raise NotImplementedError

    def reset(self) -> None:
        pass


class Sequence(Behavior):
    def __init__(self, name: str, children: list[Behavior]):
        super().__init__(name)
        self.children = children
        self.index = 0

    def tick(self, blackboard: dict) -> Status:
        while self.index < len(self.children):
            status = self.children[self.index].tick(blackboard)
            if status == Status.RUNNING:
                return Status.RUNNING
            if status == Status.FAILURE:
                self.reset()
                return Status.FAILURE
            self.children[self.index].reset()
            self.index += 1
        self.reset()
        return Status.SUCCESS

    def reset(self) -> None:
        self.index = 0
        for child in self.children:
            child.reset()


class Fallback(Behavior):
    def __init__(self, name: str, children: list[Behavior]):
        super().__init__(name)
        self.children = children
        self.index = 0

    def tick(self, blackboard: dict) -> Status:
        while self.index < len(self.children):
            status = self.children[self.index].tick(blackboard)
            if status == Status.RUNNING:
                return Status.RUNNING
            if status == Status.SUCCESS:
                self.reset()
                return Status.SUCCESS
            self.children[self.index].reset()
            self.index += 1
        self.reset()
        return Status.FAILURE

    def reset(self) -> None:
        self.index = 0
        for child in self.children:
            child.reset()


class PatrolLoop(Behavior):
    def __init__(self, child: Behavior):
        super().__init__('PatrolLoop')
        self.child = child

    def tick(self, blackboard: dict) -> Status:
        if blackboard.get('patrol_done', False):
            return Status.SUCCESS
        status = self.child.tick(blackboard)
        if status == Status.SUCCESS:
            self.child.reset()
            return Status.RUNNING
        if status == Status.FAILURE:
            if blackboard.get('patrol_done', False):
                return Status.SUCCESS
            return Status.FAILURE
        return Status.RUNNING

    def reset(self) -> None:
        self.child.reset()


class SelectWaypoint(Behavior):
    def __init__(self, navigator: BasicNavigator):
        super().__init__('SelectWaypoint')
        self.navigator = navigator

    def tick(self, blackboard: dict) -> Status:
        queue = blackboard['wp_queue']
        if not queue:
            blackboard['patrol_done'] = True
            self.navigator.get_logger().info('BT patrol complete: no waypoints left')
            return Status.FAILURE

        name, wp = queue.pop(0)
        blackboard['current_wp_name'] = name
        blackboard['current_wp'] = wp
        blackboard['navigation_failed'] = False
        total = blackboard['wp_total']
        current = total - len(queue)
        self.navigator.get_logger().info(f'[{current}/{total}] BT selected waypoint: {name}')
        return Status.SUCCESS


class NavigateToWaypoint(Behavior):
    def __init__(self, navigator: BasicNavigator, label: str,
                 stuck_timeout: float = 15.0, stuck_dist_threshold: float = 0.1):
        super().__init__(label)
        self.navigator = navigator
        self.started = False
        self.stuck_timeout = stuck_timeout
        self.stuck_dist_threshold = stuck_dist_threshold
        self.last_dist = None
        self.last_dist_time = None

    def tick(self, blackboard: dict) -> Status:
        name = blackboard['current_wp_name']
        if not self.started:
            pose = make_pose(self.navigator, blackboard['current_wp'])
            self.navigator.get_logger().info(f'{self.name}: sending Nav2 goal for {name}')
            self.navigator.goToPose(pose)
            self.started = True
            self.last_dist = None
            self.last_dist_time = None
            return Status.RUNNING

        feedback = self.navigator.getFeedback()
        now = time.time()
        if feedback and feedback.distance_remaining is not None:
            dist = feedback.distance_remaining
            if self.last_dist is not None:
                if abs(dist - self.last_dist) <= self.stuck_dist_threshold:
                    if now - self.last_dist_time >= self.stuck_timeout:
                        self.navigator.get_logger().warn(
                            f'{self.name}: STUCK detected ({self.stuck_timeout}s '
                            f'without progress). Aborting navigation.')
                        self.navigator.cancelTask()
                        blackboard['navigation_failed'] = True
                        blackboard['is_stuck'] = True
                        self.reset()
                        return Status.FAILURE
                else:
                    self.last_dist = dist
                    self.last_dist_time = now
            else:
                self.last_dist = dist
                self.last_dist_time = now

            if dist > 0:
                self.navigator.get_logger().info(f'{name}: {dist:.2f} m remaining')

        if not self.navigator.isTaskComplete():
            return Status.RUNNING

        result = self.navigator.getResult()
        if result == TaskResult.SUCCEEDED:
            self.navigator.get_logger().info(f'{self.name}: reached {name}')
            return Status.SUCCESS

        blackboard['navigation_failed'] = True
        self.navigator.get_logger().warn(f'{self.name}: failed to reach {name}')
        return Status.FAILURE

    def reset(self) -> None:
        self.started = False
        self.last_dist = None
        self.last_dist_time = None


class BackUpRecovery(Behavior):
    def __init__(self, navigator: BasicNavigator, cmd_vel_pub):
        super().__init__('BackUpRecovery')
        self.navigator = navigator
        self.cmd_vel_pub = cmd_vel_pub
        self.start_time = None

    def tick(self, blackboard: dict) -> Status:
        del blackboard
        now = time.time()
        if self.start_time is None:
            self.navigator.get_logger().warn('BT recovery: moving backwards')
            self.start_time = now

        twist = Twist()
        if now - self.start_time < 1.0:
            twist.linear.x = -0.5
            self.cmd_vel_pub.publish(twist)
            return Status.RUNNING

        self.cmd_vel_pub.publish(Twist())
        return Status.SUCCESS

    def reset(self) -> None:
        self.start_time = None


class PublishArrival(Behavior):
    def __init__(self, navigator: BasicNavigator, event_pub):
        super().__init__('PublishArrival')
        self.navigator = navigator
        self.event_pub = event_pub

    def tick(self, blackboard: dict) -> Status:
        name = blackboard['current_wp_name']
        msg = String()
        msg.data = f'llegada:{name}'
        self.event_pub.publish(msg)
        self.navigator.get_logger().info(f'BT event published: {msg.data}')
        return Status.SUCCESS


class HandleFailure(Behavior):
    def __init__(self, navigator: BasicNavigator):
        super().__init__('HandleFailure')
        self.navigator = navigator

    def tick(self, blackboard: dict) -> Status:
        name = blackboard.get('current_wp_name', 'unknown')
        if blackboard.get('skip_errors', False):
            self.navigator.get_logger().warn(f'BT skipping failed waypoint: {name}')
            return Status.SUCCESS
        self.navigator.get_logger().error(f'BT stopping on failed waypoint: {name}')
        return Status.FAILURE


def signal_handler(signum, frame):
    del signum, frame
    global stop_patrol
    stop_patrol = True


def load_waypoints() -> dict:
    try:
        pkg_share = get_package_share_directory('home_robot')
        wp_path = os.path.join(pkg_share, 'config', 'waypoints.yaml')
    except Exception:
        project_root = os.environ.get('PROJECT_ROOT', '/home/ubuntu/home_robot')
        wp_path = os.path.join(project_root, 'ros2_ws/src/home_robot/config/waypoints.yaml')

    if not os.path.exists(wp_path):
        raise FileNotFoundError(f'waypoints.yaml not found at {wp_path}')

    with open(wp_path, 'r') as stream:
        return yaml.safe_load(stream).get('waypoints', {})


def make_pose(navigator: BasicNavigator, wp: dict) -> PoseStamped:
    pose = PoseStamped()
    pose.header.frame_id = 'map'
    pose.header.stamp = navigator.get_clock().now().to_msg()
    pose.pose.position.x = float(wp['x'])
    pose.pose.position.y = float(wp['y'])
    theta = float(wp['theta'])
    pose.pose.orientation.z = math.sin(theta / 2.0)
    pose.pose.orientation.w = math.cos(theta / 2.0)
    return pose


def wait_for_nav2_action(navigator: BasicNavigator) -> bool:
    action_client = ActionClient(navigator, NavigateToPose, 'navigate_to_pose')
    start = time.time()
    while not action_client.wait_for_server(timeout_sec=1.0):
        if stop_patrol:
            return False
        if time.time() - start > 60.0:
            navigator.get_logger().error('Timeout waiting for /navigate_to_pose')
            return False
        navigator.get_logger().info('Waiting for /navigate_to_pose action server...')
    return True


def build_tree(navigator: BasicNavigator, event_pub, cmd_vel_pub) -> Behavior:
    return PatrolLoop(
        Sequence('PatrolOneWaypoint', [
            SelectWaypoint(navigator),
            Fallback('NavigateOrRecover', [
                Sequence('PrimaryNavigation', [
                    NavigateToWaypoint(navigator, 'NavigatePrimary'),
                    PublishArrival(navigator, event_pub),
                ]),
                Sequence('RecoveryNavigation', [
                    BackUpRecovery(navigator, cmd_vel_pub),
                    NavigateToWaypoint(navigator, 'NavigateRetry'),
                    PublishArrival(navigator, event_pub),
                ]),
                HandleFailure(navigator),
            ]),
        ])
    )


def main() -> None:
    global stop_patrol

    signal.signal(signal.SIGTERM, signal_handler)
    signal.signal(signal.SIGINT, signal_handler)

    use_random = '--random' in sys.argv
    skip_errors = '--skip-errors' in sys.argv

    rclpy.init()
    navigator = BasicNavigator()
    event_pub = navigator.create_publisher(String, 'patrol_events', 10)
    cmd_vel_pub = navigator.create_publisher(Twist, '/cmd_vel', 10)

    def stop_callback(msg):
        del msg
        global stop_patrol
        navigator.get_logger().warn('Stop request received on /stop_patrol')
        stop_patrol = True

    navigator.create_subscription(String, '/stop_patrol', stop_callback, 10)

    try:
        waypoints = load_waypoints()
    except FileNotFoundError as exc:
        navigator.get_logger().error(str(exc))
        rclpy.shutdown()
        return

    wp_items = list(waypoints.items())
    if use_random:
        random.shuffle(wp_items)

    blackboard = {
        'wp_queue': wp_items,
        'wp_total': len(wp_items),
        'skip_errors': skip_errors,
        'patrol_done': False,
    }

    navigator.get_logger().info('Custom BT patrol starting')
    navigator.get_logger().info(f'Mode: random={use_random}, skip_errors={skip_errors}')

    if not wait_for_nav2_action(navigator):
        rclpy.shutdown()
        return

    tree = build_tree(navigator, event_pub, cmd_vel_pub)

    try:
        while rclpy.ok() and not stop_patrol:
            rclpy.spin_once(navigator, timeout_sec=0.1)
            status = tree.tick(blackboard)
            navigator.get_logger().info(f'BT root status: {status.value}')
            if status in (Status.SUCCESS, Status.FAILURE):
                break
            time.sleep(0.2)
    finally:
        if rclpy.ok():
            navigator.cancelTask()
            cmd_vel_pub.publish(Twist())
            navigator.destroy_node()
            rclpy.shutdown()


if __name__ == '__main__':
    main()

File: ros2_ws/src/home_robot_bt/src/bt_patrol.cpp
#include <algorithm>
#include <chrono>
#include <cmath>
#include <cstdlib>
#include <memory>
#include <optional>
#include <random>
#include <string>
#include <utility>
#include <vector>

#include "ament_index_cpp/get_package_share_directory.hpp"
#include "behaviortree_cpp/bt_factory.h"
#include "behaviortree_cpp/loggers/groot2_publisher.h"
#include "geometry_msgs/msg/pose_stamped.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "nav2_msgs/action/navigate_to_pose.hpp"
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "std_msgs/msg/string.hpp"
#include "yaml-cpp/yaml.h"

using namespace std::chrono_literals;

struct Waypoint
{
  std::string name;
  double x;
  double y;
  double theta;
};

struct PatrolContext
{
  rclcpp::Node::SharedPtr node;
  rclcpp_action::Client<nav2_msgs::action::NavigateToPose>::SharedPtr nav_client;
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr event_pub;
  rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr cmd_vel_pub;
  std::vector<Waypoint> queue;
  std::optional<Waypoint> current;
  bool skip_errors{false};
};

static geometry_msgs::msg::PoseStamped makePose(
  const rclcpp::Node::SharedPtr & node, const Waypoint & wp)
{
  geometry_msgs::msg::PoseStamped pose;
  pose.header.frame_id = "map";
  pose.header.stamp = node->now();
  pose.pose.position.x = wp.x;
  pose.pose.position.y = wp.y;
  pose.pose.orientation.z = std::sin(wp.theta / 2.0);
  pose.pose.orientation.w = std::cos(wp.theta / 2.0);
  return pose;
}

static std::string waypointsPath()
{
  try {
    return ament_index_cpp::get_package_share_directory("home_robot") + "/config/waypoints.yaml";
  } catch (...) {
    const char * root = std::getenv("PROJECT_ROOT");
    const std::string project_root = root ? root : "/home/ubuntu/home_robot";
    return project_root + "/ros2_ws/src/home_robot/config/waypoints.yaml";
  }
}

static std::vector<Waypoint> loadWaypoints()
{
  YAML::Node doc = YAML::LoadFile(waypointsPath());
  std::vector<Waypoint> waypoints;
  for (const auto & item : doc["waypoints"]) {
    const auto name = item.first.as<std::string>();
    const auto wp = item.second;
    waypoints.push_back(Waypoint{
      name,
      wp["x"].as<double>(),
      wp["y"].as<double>(),
      wp["theta"].as<double>()});
  }
  return waypoints;
}

class SelectWaypoint : public BT::SyncActionNode
{
public:
  SelectWaypoint(const std::string & name, const BT::NodeConfig & config,
                 std::shared_ptr<PatrolContext> context)
  : BT::SyncActionNode(name, config), context_(std::move(context)) {}

  BT::NodeStatus tick() override
  {
    if (context_->queue.empty()) {
      RCLCPP_INFO(context_->node->get_logger(), "BT patrol complete: no waypoints left");
      return BT::NodeStatus::FAILURE;
    }
    context_->current = context_->queue.front();
    context_->queue.erase(context_->queue.begin());
    RCLCPP_INFO(
      context_->node->get_logger(), "BT selected waypoint: %s", context_->current->name.c_str());
    return BT::NodeStatus::SUCCESS;
  }

private:
  std::shared_ptr<PatrolContext> context_;
};

class NavigateToWaypoint : public BT::StatefulActionNode
{
public:
  using NavigateToPose = nav2_msgs::action::NavigateToPose;
  using GoalHandle = rclcpp_action::ClientGoalHandle<NavigateToPose>;

  NavigateToWaypoint(const std::string & name, const BT::NodeConfig & config,
                     std::shared_ptr<PatrolContext> context)
  : BT::StatefulActionNode(name, config), context_(std::move(context)) {}

  BT::NodeStatus onStart() override
  {
    if (!context_->current) {
      return BT::NodeStatus::FAILURE;
    }
    NavigateToPose::Goal goal;
    goal.pose = makePose(context_->node, *context_->current);
    RCLCPP_INFO(
      context_->node->get_logger(), "%s: sending goal to %s", name().c_str(),
      context_->current->name.c_str());
    goal_future_ = context_->nav_client->async_send_goal(goal);
    result_requested_ = false;
    return BT::NodeStatus::RUNNING;
  }

  BT::NodeStatus onRunning() override
  {
    if (!goal_handle_) {
      if (goal_future_.valid() && goal_future_.wait_for(0ms) == std::future_status::ready) {
        goal_handle_ = goal_future_.get();
        if (!goal_handle_) {
          RCLCPP_WARN(context_->node->get_logger(), "%s: goal rejected", name().c_str());
          return BT::NodeStatus::FAILURE;
        }
      } else {
        return BT::NodeStatus::RUNNING;
      }
    }

    if (!result_requested_) {
      result_future_ = context_->nav_client->async_get_result(goal_handle_);
      result_requested_ = true;
      return BT::NodeStatus::RUNNING;
    }

    if (result_future_.valid() && result_future_.wait_for(0ms) == std::future_status::ready) {
      const auto result = result_future_.get();
      goal_handle_.reset();
      result_requested_ = false;
      if (result.code == rclcpp_action::ResultCode::SUCCEEDED) {
        RCLCPP_INFO(context_->node->get_logger(), "%s: navigation succeeded", name().c_str());
        return BT::NodeStatus::SUCCESS;
      }
      RCLCPP_WARN(context_->node->get_logger(), "%s: navigation failed", name().c_str());
      return BT::NodeStatus::FAILURE;
    }

    return BT::NodeStatus::RUNNING;
  }

  void onHalted() override
  {
    if (goal_handle_) {
      context_->nav_client->async_cancel_goal(goal_handle_);
    }
    goal_handle_.reset();
    result_requested_ = false;
  }

private:
  std::shared_ptr<PatrolContext> context_;
  std::shared_future<GoalHandle::SharedPtr> goal_future_;
  GoalHandle::SharedPtr goal_handle_;
  std::shared_future<GoalHandle::WrappedResult> result_future_;
  bool result_requested_{false};
};

class BackUpRecovery : public BT::StatefulActionNode
{
public:
  BackUpRecovery(const std::string & name, const BT::NodeConfig & config,
                 std::shared_ptr<PatrolContext> context)
  : BT::StatefulActionNode(name, config), context_(std::move(context)) {}

  BT::NodeStatus onStart() override
  {
    start_ = context_->node->now();
    RCLCPP_WARN(context_->node->get_logger(), "BT recovery: moving backwards");
    return BT::NodeStatus::RUNNING;
  }

  BT::NodeStatus onRunning() override
  {
    geometry_msgs::msg::Twist twist;
    if ((context_->node->now() - start_).seconds() < 1.2) {
      twist.linear.x = -0.25;
      context_->cmd_vel_pub->publish(twist);
      return BT::NodeStatus::RUNNING;
    }
    context_->cmd_vel_pub->publish(twist);
    return BT::NodeStatus::SUCCESS;
  }

  void onHalted() override
  {
    context_->cmd_vel_pub->publish(geometry_msgs::msg::Twist());
  }

private:
  std::shared_ptr<PatrolContext> context_;
  rclcpp::Time start_;
};

class PublishArrival : public BT::SyncActionNode
{
public:
  PublishArrival(const std::string & name, const BT::NodeConfig & config,
                 std::shared_ptr<PatrolContext> context)
  : BT::SyncActionNode(name, config), context_(std::move(context)) {}

  BT::NodeStatus tick() override
  {
    if (!context_->current) {
      return BT::NodeStatus::FAILURE;
    }
    std_msgs::msg::String msg;
    msg.data = "llegada:" + context_->current->name;
    context_->event_pub->publish(msg);
    RCLCPP_INFO(context_->node->get_logger(), "BT event published: %s", msg.data.c_str());
    return BT::NodeStatus::SUCCESS;
  }

private:
  std::shared_ptr<PatrolContext> context_;
};

class HandleFailure : public BT::SyncActionNode
{
public:
  HandleFailure(const std::string & name, const BT::NodeConfig & config,
                std::shared_ptr<PatrolContext> context)
  : BT::SyncActionNode(name, config), context_(std::move(context)) {}

  BT::NodeStatus tick() override
  {
    const auto wp = context_->current ? context_->current->name : "unknown";
    if (context_->skip_errors) {
      RCLCPP_WARN(context_->node->get_logger(), "BT skipping failed waypoint: %s", wp.c_str());
      return BT::NodeStatus::SUCCESS;
    }
    RCLCPP_ERROR(context_->node->get_logger(), "BT stopping on failed waypoint: %s", wp.c_str());
    return BT::NodeStatus::FAILURE;
  }

private:
  std::shared_ptr<PatrolContext> context_;
};

class PatrolLoop : public BT::DecoratorNode
{
public:
  PatrolLoop(const std::string & name, const BT::NodeConfig & config,
             std::shared_ptr<PatrolContext> context)
  : BT::DecoratorNode(name, config), context_(std::move(context)) {}

  BT::NodeStatus tick() override
  {
    setStatus(BT::NodeStatus::RUNNING);
    const auto child_status = child_node_->executeTick();
    if (child_status == BT::NodeStatus::SUCCESS) {
      haltChild();
      return BT::NodeStatus::RUNNING;
    }
    if (child_status == BT::NodeStatus::FAILURE && context_->queue.empty()) {
      haltChild();
      return BT::NodeStatus::SUCCESS;
    }
    return child_status;
  }

private:
  std::shared_ptr<PatrolContext> context_;
};

static bool hasArg(int argc, char ** argv, const std::string & arg)
{
  for (int i = 1; i < argc; ++i) {
    if (argv[i] == arg) {
      return true;
    }
  }
  return false;
}

static std::string valueArg(int argc, char ** argv, const std::string & prefix, const std::string & fallback)
{
  for (int i = 1; i < argc; ++i) {
    const std::string arg = argv[i];
    if (arg.rfind(prefix, 0) == 0) {
      return arg.substr(prefix.size());
    }
  }
  return fallback;
}

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);

  const bool random_order = hasArg(argc, argv, "--random");
  const bool skip_errors = hasArg(argc, argv, "--skip-errors");
  const auto default_xml = ament_index_cpp::get_package_share_directory("home_robot_bt") +
    "/config/bt_patrol.xml";
  auto bt_xml = valueArg(argc, argv, "--bt-xml=", default_xml);
  if (bt_xml.empty()) {
    bt_xml = default_xml;
  }
  const int groot_port = std::stoi(valueArg(argc, argv, "--groot-port=", "1667"));

  auto context = std::make_shared<PatrolContext>();
  context->node = rclcpp::Node::make_shared("bt_patrol");
  context->nav_client = rclcpp_action::create_client<nav2_msgs::action::NavigateToPose>(
    context->node, "navigate_to_pose");
  context->event_pub = context->node->create_publisher<std_msgs::msg::String>("patrol_events", 10);
  context->cmd_vel_pub = context->node->create_publisher<geometry_msgs::msg::Twist>("cmd_vel", 10);
  context->skip_errors = skip_errors;
  context->queue = loadWaypoints();

  if (random_order) {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::shuffle(context->queue.begin(), context->queue.end(), gen);
  }

  RCLCPP_INFO(context->node->get_logger(), "Custom BehaviorTree.CPP patrol starting");
  RCLCPP_INFO(context->node->get_logger(), "BT XML: %s", bt_xml.c_str());
  RCLCPP_INFO(context->node->get_logger(), "Groot2 live publisher port: %d", groot_port);

  BT::BehaviorTreeFactory factory;
  factory.registerBuilder<SelectWaypoint>(
    "SelectWaypoint", [context](const std::string & name, const BT::NodeConfig & config) {
      return std::make_unique<SelectWaypoint>(name, config, context);
    });
  factory.registerBuilder<NavigateToWaypoint>(
    "NavigateToWaypoint", [context](const std::string & name, const BT::NodeConfig & config) {
      return std::make_unique<NavigateToWaypoint>(name, config, context);
    });
  factory.registerBuilder<BackUpRecovery>(
    "BackUpRecovery", [context](const std::string & name, const BT::NodeConfig & config) {
      return std::make_unique<BackUpRecovery>(name, config, context);
    });
  factory.registerBuilder<PublishArrival>(
    "PublishArrival", [context](const std::string & name, const BT::NodeConfig & config) {
      return std::make_unique<PublishArrival>(name, config, context);
    });
  factory.registerBuilder<HandleFailure>(
    "HandleFailure", [context](const std::string & name, const BT::NodeConfig & config) {
      return std::make_unique<HandleFailure>(name, config, context);
    });
  factory.registerBuilder<PatrolLoop>(
    "PatrolLoop", [context](const std::string & name, const BT::NodeConfig & config) {
      return std::make_unique<PatrolLoop>(name, config, context);
    });

  auto tree = factory.createTreeFromFile(bt_xml);
  std::unique_ptr<BT::Groot2Publisher> groot_publisher;
  try {
    groot_publisher = std::make_unique<BT::Groot2Publisher>(
      tree, static_cast<unsigned>(groot_port));
  } catch (const std::exception & ex) {
    RCLCPP_WARN(
      context->node->get_logger(), "Could not start Groot2 publisher on port %d: %s",
      groot_port, ex.what());
  }

  if (!context->nav_client->wait_for_action_server(60s)) {
    RCLCPP_ERROR(context->node->get_logger(), "Timeout waiting for navigate_to_pose action server");
    rclcpp::shutdown();
    return 1;
  }

  rclcpp::Rate rate(5.0);
  BT::NodeStatus status = BT::NodeStatus::IDLE;
  while (rclcpp::ok()) {
    rclcpp::spin_some(context->node);
    status = tree.tickOnce();
    if (status == BT::NodeStatus::SUCCESS || status == BT::NodeStatus::FAILURE) {
      break;
    }
    rate.sleep();
  }

  tree.haltTree();
  context->cmd_vel_pub->publish(geometry_msgs::msg::Twist());
  RCLCPP_INFO(context->node->get_logger(), "BT finished with status: %s", BT::toStr(status).c_str());
  rclcpp::shutdown();
  return status == BT::NodeStatus::FAILURE ? 1 : 0;
}
